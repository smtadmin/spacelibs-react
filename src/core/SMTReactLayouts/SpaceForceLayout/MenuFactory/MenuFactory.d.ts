export default MenuFactory;
declare class MenuFactory {
    constructor(props: any);
    props: any;
    buildNavLinks(): void;
    menuElements: any[];
    content: JSX.Element;
    render(): JSX.Element;
}
declare namespace MenuFactory {
    namespace defaultProps {
        const text: boolean;
        const icon: boolean;
        const dropdown: boolean;
    }
    namespace propTypes {
        export const test: any;
        const icon_1: any;
        export { icon_1 as icon };
        export const name: any;
        const dropdown_1: any;
        export { dropdown_1 as dropdown };
    }
}
