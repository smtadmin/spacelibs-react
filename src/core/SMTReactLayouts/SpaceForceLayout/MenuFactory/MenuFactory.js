/*
 * File: MenuFactory
 * Version: 0.1.0
 * Project: react-test
 * Description: Takes a route data object and builds out a menu element based on the set parameters
 * By default returns both the icon and the link name but can be overridden by setting the icon or the text props to false
 * A name props should be added to increase the specificity of the generated class tags
 * File Created: Friday, 5th February 2021 10:31 am
 * Author: Justin Jeffrey (justin.jeffrey@siliconmtn.com)
 * -----
 * Last Modified: Wednesday, 10th February 2021 3:02 pm
 * Modified By: Justin Jeffrey (justin.jeffrey@siliconmtn.com>)
 * -----
 * Copyright 2021, Silicon Mountain Technologies, Inc.
 */
import React from 'react';
import { Nav, NavDropdown } from 'react-bootstrap';
import PropTypes from 'prop-types';

/**
 * Takes a route data object and builds out a menu element
 */
class MenuFactory extends React.Component {

    /**
     * Creates an instance of the MenuFactory Class
     * @param {*} props 
     */
    constructor(props) {
        super();
        this.props = props;
        this.buildNavLinks(props.routeData);
    }

    /**
     * Builds a list of Nav.Link elements using the array of objects passed by the routeData param
     * @param {*} routeData 
     */
    buildNavLinks(routeData) {

        this.menuElements = [];
        let count = 0;
        routeData.forEach(route => {
            let element = [];
            if (this.props.icon === true) {
                element.push(<div key={"icon-" + count} className={"px-1 nav-icon nav-element-icon-" + count} >{route.icon}</div>);
            }
            if (this.props.text === true) {
                element.push(<div key={"text-" + count} className={"px-1 nav-text nav-element-text-" + count} >{route.text}</div>);
            }
            if (this.props.dropdown) {
                this.menuElements.push(<NavDropdown.Item key={this.props.name + count} className={this.props.name + "-nav-element-" + count + " d-flex"} href={route.link}>{element}</NavDropdown.Item>)
                this.content = <NavDropdown title={this.props.name} className={this.props.name}>{this.menuElements}</NavDropdown>;
            } else {
                this.menuElements.push(<Nav.Link key={this.props.name + count} className={this.props.name + "-nav-element-" + count + " d-flex"} href={route.link}>{element}</Nav.Link>)
                this.content = <Nav title="nav-menu" className={this.props.name}>{this.menuElements}</Nav>;
            }
            count++;
        });

    }

    /**
     * Returns a menu component
     */
    render() {
        return this.content;
    }

}

/** Sets Default Props */
MenuFactory.defaultProps = {
    text: true,
    icon: true,
    dropdown: false
};
/** Prop Validators */
MenuFactory.propTypes = {
    test: PropTypes.bool,
    icon: PropTypes.bool,
    name: PropTypes.string.isRequired,
    dropdown: PropTypes.bool
};

export default MenuFactory;