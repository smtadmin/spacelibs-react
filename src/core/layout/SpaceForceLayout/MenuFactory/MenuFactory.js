/*
 * File: MenuFactory
 * Version: 0.1.0
 * Project: react-test
 * Description: Takes a route data object and builds out a menu element based on the set parameters
 * By default returns both the icon and the link name but can be overridden by setting the icon or the text props to false
 * A name props should be added to increase the specificity of the generated class tags
 * File Created: Friday, 5th February 2021 10:31 am
 * Author: Justin Jeffrey (justin.jeffrey@siliconmtn.com)
 * -----
 * Last Modified: Wednesday, 3rd March 2021 5:07 pm
 * Modified By: Justin Jeffrey (justin.jeffrey@siliconmtn.com>)
 * -----
 * Copyright 2021, Silicon Mountain Technologies, Inc.
 */
import React from 'react';
import Nav from 'react-bootstrap/Nav';
import NavDropdown from 'react-bootstrap/NavDropdown';
import PropTypes from 'prop-types';

/**
 * Takes a route data object and builds out a menu element
 */
class MenuFactory extends React.Component {

    /**
     * Creates an instance of the MenuFactory Class
     * @param {*} props 
     */
    constructor(props) {
        super();
        this.props = props;
        this.buildNavLinks();
    }

    /**
     * Builds a list of Nav.Link elements using the array of objects passed by the routeData param
     */
    buildNavLinks() {

        this.menuElements = [];
        let count = 0;
        const routeData = this.props.routeData;
        Object.keys(routeData).forEach(function (item) {
            let element = [];
            if (this.props.icon === true) {
                element.push(<div key={"icon-" + count} className={"px-1 nav-icon nav-element-icon-" + count} >{routeData[item].icon}</div>);
            }
            if (this.props.text === true) {
                element.push(<div key={"text-" + count} className={"px-1 nav-text nav-element-text-" + count} >{routeData[item].text}</div>);
            }
            if (this.props.dropdown) {
                this.menuElements.push(<NavDropdown.Item key={routeData[item].name + count} className={routeData[item].name + "-nav-element-" + count + " d-flex"} href={routeData[item].link}>{element}</NavDropdown.Item>)
                this.content = <NavDropdown title={this.props.name} className={routeData[item].name}>{this.menuElements}</NavDropdown>;
            } else {
                this.menuElements.push(<Nav.Link key={routeData[item].name + count} className={routeData[item].name + "-nav-element-" + count + " d-flex"} href={routeData[item].link}>{element}</Nav.Link>)
                this.content = <Nav title="nav-menu" className={routeData[item].name}>{this.menuElements}</Nav>;
            }
            count++;
        }, this);

    }

    /**
     * Returns a menu component
     */
    render() {
        return this.content;
    }

}

/** Sets Default Props */
MenuFactory.defaultProps = {
    text: true,
    icon: true,
    dropdown: false
};
/** Prop Validators */
MenuFactory.propTypes = {
    text: PropTypes.bool,
    icon: PropTypes.bool,
    name: PropTypes.string.isRequired,
    dropdown: PropTypes.bool
};

export default MenuFactory;