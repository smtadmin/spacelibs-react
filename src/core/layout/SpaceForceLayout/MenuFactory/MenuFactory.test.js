/*
 * File: MenuFactory.test
 * Version: 0.1.0
 * Project: react-test
 * Description: Testing suite for the MenuFactory Component
 * File Created: Monday, 8th February 2021 12:19 pm
 * Author: Justin Jeffrey (justin.jeffrey@siliconmtn.com)
 * -----
 * Last Modified: Wednesday, 10th February 2021 3:14 pm
 * Modified By: Justin Jeffrey (justin.jeffrey@siliconmtn.com>)
 * -----
 * Copyright 2021, Silicon Mountain Technologies, Inc.
 */
import MenuFactory from './MenuFactory';
import React from 'react';
import { render } from '@testing-library/react';


const demoRoutes = {
    publicRouteData: [
        {
            text: "Analytics",
            icon: "Analytics Icon",
            link: "/analytics",
            children: []
        }
    ],
    adminRouteData: [
        {
            text: "Admin Page",
            icon: "admin Page Icon",
            link: "/admin",
            children: []
        }
    ]
};

/**
 * Attempts to render a MenuFactory component to confirm that it renders without crashing
 */
it("Renders without crashing", () => {

    const element = render(<MenuFactory name="mobile-public-menu" routeData={demoRoutes.publicRouteData} />);
    expect(element).toBeTruthy();

});

/**
 * Renders a MenuFactory component using the publicRouteData object and compares the render output to a snapshot
 */
it("Renders Public Menus Correctly", () => {

    const element = render(<MenuFactory name="mobile-public-menu" routeData={demoRoutes.publicRouteData} />);
    expect(element).toMatchSnapshot();
})

/**
 * Renders a MenuFactory component using the adminRouteData object  and compares the render output to a snapshot
 */
it("Renders Admin Menus Correctly", () => {

    const element = render(<MenuFactory name="mobile-public-menu" routeData={demoRoutes.adminRouteData} />);
    expect(element).toMatchSnapshot();
})

/**
 * Renders a MenuFactory component using the publicRouteData object with the icon prop set to false and compares the render output to a snapshot
 */
it("Renders Public Menus Correctly Without Icons", () => {

    const element = render(<MenuFactory name="mobile-public-menu" icon={false} routeData={demoRoutes.publicRouteData} />);
    expect(element).toMatchSnapshot();
})

/**
 * Renders a MenuFactory component using the publicRouteData object with the dropdown prop set to true and compares the render output to a snapshot
 */
it("Renders A Dropdown Menu Correctly", () => {

    const element = render(<MenuFactory name="dropdown-test" routeData={demoRoutes.publicRouteData} dropdown={true} />);
    expect(element).toMatchSnapshot();
    
})

/**
 * Renders a MenuFactory component using the publicRouteData object with the text prop set to false and compares the render output to a snapshot
 */
it("Renders Public Menus Correctly Without Text", () => {

    const element = render(<MenuFactory name="mobile-public-menu" text={false} routeData={demoRoutes.publicRouteData} />);
    expect(element).toMatchSnapshot();
    
})