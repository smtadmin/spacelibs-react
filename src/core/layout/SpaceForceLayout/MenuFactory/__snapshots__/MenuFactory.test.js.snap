// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders A Dropdown Menu Correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="nav-item dropdown"
      >
        <a
          aria-expanded="false"
          aria-haspopup="true"
          class="dropdown-toggle nav-link"
          href="#"
          role="button"
        >
          dropdown-test
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="nav-item dropdown"
    >
      <a
        aria-expanded="false"
        aria-haspopup="true"
        class="dropdown-toggle nav-link"
        href="#"
        role="button"
      >
        dropdown-test
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Renders Admin Menus Correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="nav"
        title="nav-menu"
      >
        <a
          class="undefined-nav-element-0 d-flex nav-link"
          data-rb-event-key="/admin"
          href="/admin"
        >
          <div
            class="px-1 nav-icon nav-element-icon-0"
          >
            admin Page Icon
          </div>
          <div
            class="px-1 nav-text nav-element-text-0"
          >
            Admin Page
          </div>
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="nav"
      title="nav-menu"
    >
      <a
        class="undefined-nav-element-0 d-flex nav-link"
        data-rb-event-key="/admin"
        href="/admin"
      >
        <div
          class="px-1 nav-icon nav-element-icon-0"
        >
          admin Page Icon
        </div>
        <div
          class="px-1 nav-text nav-element-text-0"
        >
          Admin Page
        </div>
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Renders Public Menus Correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="nav"
        title="nav-menu"
      >
        <a
          class="undefined-nav-element-0 d-flex nav-link"
          data-rb-event-key="/analytics"
          href="/analytics"
        >
          <div
            class="px-1 nav-icon nav-element-icon-0"
          >
            Analytics Icon
          </div>
          <div
            class="px-1 nav-text nav-element-text-0"
          >
            Analytics
          </div>
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="nav"
      title="nav-menu"
    >
      <a
        class="undefined-nav-element-0 d-flex nav-link"
        data-rb-event-key="/analytics"
        href="/analytics"
      >
        <div
          class="px-1 nav-icon nav-element-icon-0"
        >
          Analytics Icon
        </div>
        <div
          class="px-1 nav-text nav-element-text-0"
        >
          Analytics
        </div>
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Renders Public Menus Correctly Without Icons 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="nav"
        title="nav-menu"
      >
        <a
          class="undefined-nav-element-0 d-flex nav-link"
          data-rb-event-key="/analytics"
          href="/analytics"
        >
          <div
            class="px-1 nav-text nav-element-text-0"
          >
            Analytics
          </div>
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="nav"
      title="nav-menu"
    >
      <a
        class="undefined-nav-element-0 d-flex nav-link"
        data-rb-event-key="/analytics"
        href="/analytics"
      >
        <div
          class="px-1 nav-text nav-element-text-0"
        >
          Analytics
        </div>
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Renders Public Menus Correctly Without Text 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="nav"
        title="nav-menu"
      >
        <a
          class="undefined-nav-element-0 d-flex nav-link"
          data-rb-event-key="/analytics"
          href="/analytics"
        >
          <div
            class="px-1 nav-icon nav-element-icon-0"
          >
            Analytics Icon
          </div>
        </a>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="nav"
      title="nav-menu"
    >
      <a
        class="undefined-nav-element-0 d-flex nav-link"
        data-rb-event-key="/analytics"
        href="/analytics"
      >
        <div
          class="px-1 nav-icon nav-element-icon-0"
        >
          Analytics Icon
        </div>
      </a>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
